rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,205 @@
+
+    private @MonotonicNonNull ClientCapabilities clientCapabilities;
+    private @MonotonicNonNull ServerCapabilities serverCapabilities;
+
+    public DynamicCapabilities(LanguageClient client, Executor exec, List<AbstractDynamicCapability<?>> supportedCapabilities) {
+        this.client = client;
+        this.exec = exec;
+        this.supportedCapabilities = supportedCapabilities;
+        this.staticCapabilities = new HashSet<>();
+    }
+
+    /**
+     * Determines whether the server should register capabiltities statically now instead of dynamically later.
+     * @param clientCapabilities The capabilities of the client, which determine whether dynamic registration is supported at all.
+     * @param serverCapabilities The server capabilities to modify.
+     */
+    public void setStaticCapabilities(ClientCapabilities clientCapabilities, final ServerCapabilities serverCapabilities) {
+        // Use these to determine whether we actually need/are allowed to register certain capabilities
+        // If the client does not support dynamic registration for a certain capability, do not register it. Instead, register it statically.
+        this.clientCapabilities = clientCapabilities;
+        this.serverCapabilities = serverCapabilities;
+
+        for (var cap : supportedCapabilities) {
+            if (!cap.checkDynamicCapability(clientCapabilities, serverCapabilities)) {
+                staticCapabilities.add(cap);
+            }
+        }
+    }
+
+    /**
+     * Update capabilities for language contributions.
+     * @param contribs The contributions to represent
+     * @return A void future that completes when all capabilities are updated.
+     */
+    public CompletableFuture<Void> updateCapabilities(Map<String,LanguageContributionsMultiplexer> contributions) {
+        return updateCapabilities(contributions.values().stream().map(ILanguageContributions.class::cast).collect(Collectors.toList()));
+    }
+
+    /**
+     * Update capabilities for language contributions.
+     * @param contribs The contributions to represent.
+     * @return A void future that completes when all capabilities are updated.
+     */
+    CompletableFuture<Void> updateCapabilities(Collection<ILanguageContributions> contribs) {
+        // Compute registrations purely based on contributions
+        // This requires waiting for an evaluator to load, which might take long, and should not block our logbook
+        return CompletableFutureUtils.reduce(supportedCapabilities.stream().filter(cap -> !staticCapabilities.contains(cap)).map(c -> maybeRegistration(c, contribs)), exec)
+            .thenAccept(capabilities -> {
+                // Since we have some bookkeeping to do, we will now block our logbook for a moment
+                synchronized (currentRegistrations) {
+                    List<Registration> registrations = new LinkedList<>();
+                    List<Unregistration> unregistrations = new LinkedList<>();
```

Comment:
I'm not convinced that this code will not cause a race. The combination of all these futures with a lock on the `currentRegistrations` smells like deadlock waiting to happen. For example with `doRegistrations` that locks on the `this`.

I think it would work better if we change the code to first build the `registration` and `unregistration` work lists. and then manipulate the `currentRegistrations` based on that, without a lock, but with the remove/puts that validate the current state in the map. and based on that get a queue of things to actually send to the server.